## Data Preprocessing
1. Use `cutadapt` to trim adapter sequences and discard reads containing ambiguous 'N' bases (`cutadapt --max-n 0 -a AGATCGGAAGAGCACACGTCT`)
2. Collapse duplicate reads to reduce redundancy, calculate the read count for each unique read and output the result in FASTA format
3. calculating normalization factor
    - miRNA abundance: Calculate the read count of perfectly mapped reads to known miRNAs(using `bowtie2`), then normalize `normalization factor = (mapped_miRNA_read_count) / 10^6`
    - RPM: Sum up the read counts of all collapsed reads, `normalization_factor = (total_read_count) / 10^6`
4. Filter reads that start with a 'G', have a length of 22–23 nucleotides and map filtered reads to the mRNA reference using bowtie2 in perfect-match mode (`bowtie2 -a --score-min C,0,0 --norc --no-unal --no-hd`).
5. output .csv file

---

## Data Download

Please download the required normalization data files from the following links:

- [RPM normalization data](http://nas.csblab.ee.ncku.edu.tw:32200/sharing/pMyHlDuA8)  
- [miRNA abundance normalization data](http://nas.csblab.ee.ncku.edu.tw:32200/sharing/IaGDgQxT3)  

After downloading, place the files into the following directory: 
`RHA-1_generate_fig/input/`

Example:
If you only need to analyze data normalized using RPM, you just need to download the CSV file from [this link](http://nas.csblab.ee.ncku.edu.tw:32200/sharing/pMyHlDuA8) and place it in the input folder. There's no need to download the other CSV file from the second link.

---

## Required Packages and System Dependencies

This program requires specific Python packages and system-level dependencies to run correctly. Please ensure that all versions match those listed below to maintain compatibility.

### Python Package Requirements

Install the following Python packages with the specified versions:

```bash
pip install \
  numpy==1.18.5 \
  pandas==0.24.2 \
  matplotlib==3.0.3 \
  seaborn==0.9.1 \
  pysam==0.20.0 \
  oyaml==1.0 \
  scipy==1.4.1 \
  statannotations==0.6.0 \
  rpy2==3.0.5 \
  tzlocal==2.1 \
  cutadapt==3.7 \
  PyYAML==5.3.1 \
  tqdm==4.64.1
```

### System-Level Requirements
- Python version: `3.6`
- System package: 
    <!-- - cutadapt version `2.9`
    ```bash
    apt install cutadapt  
    ``` -->
    - bowtie2 version `2.5.3`
    ```bash
    cd path-to-RHA-1_pipeline/
    wget https://sourceforge.net/projects/bowtie-bio/files/bowtie2/2.5.3/bowtie2-2.5.3-linux-x86_64.zip/download -O Bowtie2.zip
    unzip -j Bowtie2.zip -d Bowtie2
    mv Bowtie2 sRNAanalyst/src/
    rm Bowtie2.zip
    ```
    <!-- add the line below to 
    ```bash
    export PATH=sRNAanalyst/src/Bowtie2:$PATH
    ``` -->
    - other package
    ```bash
    apt install -y libbz2-dev liblzma-dev libcurl4-openssl-dev libssl-dev libncurses5-dev build-essential unzip
    ```
---

---
## Working with `.fastq` Files
If you have .fastq.gz files, please include the following samples in the input directory: 
`OG-02, OG-06, OG-10, OG-14, OG-04, OG-08, OG-12, OG-16, OG-03, OG-07` `RHA-1_generate_fig/input/` 

Rename them appropriately, for example: `OG-01.fastq.gz`

Then, modify the `RHA-1_generate_fig/run_generate_fig.sh` script by changing: 
```diff
- preprocess=False
+ preprocess=True
```

Users can customize the normalization method by modifying the following line in the script `RHA-1_generate_fig/run_generate_fig.sh`:
```
# [ "RPM" / "miRNA_abu" ]
norm_method="RPM"
```

## Executing Program
After building the virtual environment, and put csv files in `RHA-1_generate_fig/input/`
```bash
cd path-to-RHA-1_generate_fig/
bash run_generate_fig.sh
```

After executing bash `run_generate_fig.sh`, all figures from the paper will be saved in the `RHA-1_generate_fig/output/` directory.

---
## Example Outputs
This folder contains example images generated by the pipeline.  
To regenerate them, run `bash run_generate_fig.sh` with appropriate input data.

- The upper image(`fig_e1.png`) corresponds to the left scatter plot of Figure E, and the bottom image(`fig_e2.png`) corresponds to the right scatter plot of Figure E.
![fig_e1](./example_output/fig_e1.png)
![fig_e2](./example_output/fig_e2.png)

- The upper image(`fig_f1.png`) corresponds to the left scatter plot of Figure F, and the bottom image(`fig_f2.png`) corresponds to the right scatter plot of Figure F.
![fig_f1](./example_output/fig_f1.png)
![fig_f2](./example_output/fig_f2.png)

- The bottom plot in this image(`fig_g.png`) represents Figure G in the paper.
![fig_g](./example_output/fig_g.png)

- The bottom plot in this image(`fig_h.png`) represents Figure H in the paper.
![fig_h](./example_output/fig_h.png)

- This is the bedGraph(`OG-03_22G.png`) of OG-03. After running `run_generate_fig.sh`, bedGraph files for OG-03, OG-04, OG-07, OG-08, OG-12, and OG-16 are generated and saved in the `output` folder.
![OG-03_22G](./example_output/OG-03_22G.png)

---
## External Code Reference

This project includes adapted code from the [sRNAanalyst](https://github.com/RyanCCJ/sRNAanalyst) project by [RyanCCJ](https://github.com/RyanCCJ), specifically for small RNA analysis functionalities.

Only selected scripts or modules have been retained under `RHA-1_generate_fig/sRNAanalyst/`, and modifications may have been made for compatibility or customization.

> Original repository: https://github.com/RyanCCJ/sRNAanalyst  
> License: MIT License – original license and attribution have been preserved where applicable.

<!-- ## Svg/Png mode transfer
for example png -> svg mode
- `run_generate_fig.sh`
```diff
- mv output/analyze/fig/Metagene_0.png ../../output/fig_h.png
+ mv output/analyze/fig/Metagene_0.svg ../../output/fig_h.svg
```
```diff
- mv output/analyze/fig/Metagene_0.png ../../output/fig_g.png
+ mv output/analyze/fig/Metagene_0.svg ../../output/fig_g.svg
```
- `sRNAanalyst/example/config/stylesheet.yml`
```diff
- fig_format: png
+ fig_format: svg
```
- `code/process_22G_bedgraph.py`
```diff
- output_filename = "output/{}_22G.png".format(os.path.basename(csv_file).replace(".csv", ''))
+ output_filename = "output/{}_22G.svg".format(os.path.basename(csv_file).replace(".csv", ''))
```
- `code/run_scatter.py`
```diff
- os.system("mv output/analyze/fig/Scatter_0.png ../../output/fig_f1.png")
+ os.system("mv output/analyze/fig/Scatter_0.svg ../../output/fig_f1.svg")
```
```diff
- os.system("mv output/analyze/fig/Scatter_0.png ../../output/fig_e1.png")
+ os.system("mv output/analyze/fig/Scatter_0.svg ../../output/fig_e1.svg")
```
```diff
- os.system("mv output/analyze/fig/Scatter_0.png ../../output/fig_f2.png")
+ os.system("mv output/analyze/fig/Scatter_0.svg ../../output/fig_f2.svg")
```
```diff
- os.system("mv output/analyze/fig/Scatter_0.png ../../output/fig_e2.png")
+ os.system("mv output/analyze/fig/Scatter_0.svg ../../output/fig_e2.svg")
``` -->
